File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = defaultdict(int)
        count = i = 0
        for ch in s:
            chars[ch] += 1
            count += 1
            if len(chars) < count:
                count -= 1
                chars[s[i]] -= 1
                if chars[s[i]] == 0:
                    del chars[s[i]]
                i += 1
        return len(s) - i

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        
        unordered_map < char, int> uniques;
        int start = -1, res = 0, len = s.length();
        for (int i = 0; i < len; ++i){
            if (uniques.find(s[i]) != uniques.end() && uniques[s[i]] > start){
                start = uniques[s[i]];
            }
            uniques[s[i]] = i;
            if (res < i - start){
                res = i - start;
            }
        }
        return res;
    }
};